Python/specialize.c

```diff
@@ -945,24 +945,34 @@ specialize_class_load_attr(PyObject *owner, _Py_CODEUNIT *instr,
                              PyObject *name)
 {
     _PyLoadMethodCache *cache = (_PyLoadMethodCache *)(instr + 1);
+    PyTypeObject *metaclass = Py_TYPE(owner);
+    if (_PyType_Lookup(metaclass, name)) {
+        // The metaclass has the named attribute. This breaks LOAD_ATTR_CLASS
+        // when the attribute *also* exists on the class... but it also means we
+        // can't specialize for things like Class.__doc__, Class.__name__, etc.
+        // Perhaps LOAD_ATTR_METACLASS may be worth adding in the future?
+        SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_ATTR_METACLASS_ATTRIBUTE);
+        return -1;
+    }
+    if (metaclass->tp_getattro != PyType_Type.tp_getattro) {
+        // The metaclass defines __getattribute__. All bets are off:
+        SPECIALIZATION_FAIL(LOAD_ATTR, SPEC_FAIL_OVERRIDDEN);
+        return -1;
+    }
```

Comment:
Do you think we should just play safe and fail whenever we see a non-`type` type? (ie whenever there's a metaclass)

Reply:
Thanks for the quick work @brandtbucher 

I don't know if this helps but if you want to time SymPy specifically then one way is:
```
import sympy
sympy.test()
```
That takes about half an hour to run on this computer.

Comparing different CPython versions in CI shows that 3.11.0b4 currently gives a speedup for some jobs and is neutral for others:
https://github.com/sympy/sympy/actions/runs/2662492934

If you look at for example the `test2` matrix then the timings from there are:
```
3.8: 19m28s
3.9: 16m44s
3.10: 16m4s
3.11.0b4: 14m45s
```
The test2 job does roughly half of what `sympy.test()` does and can be reproduce directly with `sympy.test(split='2/2')`. The 3.10 timing is shown in CI as `test2-latest` separately from the `test2` matrix of different Python versions.

Reply:
Thanks, I'll take a look.

Side note, I just realized that potentially removing support for *any* metaclasses here would probably (further) trash `enum` performance. :( So that's also something we want visibility into before going that route.